mmdet3d/ops/paconv/paconv.py

```diff
@@ -0,0 +1,350 @@
+                 weight_bank_init: str = 'kaiming',
+                 kernel_input: str = 'neighbor',
+                 scorenet_cfg: dict = dict(
+                     mlp_channels=[8, 16, 16],
+                     score_norm='softmax',
+                     temp_factor=1.0,
+                     last_bn=False)):
+        super(PAConvCUDA, self).__init__(
+            in_channels=in_channels,
+            out_channels=out_channels,
+            num_kernels=num_kernels,
+            norm_cfg=norm_cfg,
+            act_cfg=act_cfg,
+            scorenet_input=scorenet_input,
+            weight_bank_init=weight_bank_init,
+            kernel_input=kernel_input,
+            scorenet_cfg=scorenet_cfg)
+
+        assert self.kernel_input == 'neighbor', \
+            'CUDA implemented PAConv only supports neighbor kernel_input'
+
+    def forward(
+        self, points_xyz: torch.Tensor, features: torch.Tensor,
+        points_idx: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Forward.
+
+        Args:
+            points_xyz (torch.Tensor): (B, 3, npoint, K)
+                Coordinates of the grouped points.
+            features (torch.Tensor): (B, in_c, N)
+                Features of all points in the current point cloud.
+                Different from `features` in non-CUDA version PAConv, here the
+                    features are not grouped by each center to form a K dim.
+            points_idx (torch.Tensor): (B, npoint, K)
+                Index of the grouped points.
+
+        Returns:
+            torch.Tensor: (B, out_c, npoint, K), features after PAConv.
+        """
+        B, _, npoint, K = points_xyz.size()
+        center_xyz = points_xyz[..., :1].repeat(1, 1, 1, K)
+        xyz_diff = points_xyz - center_xyz  # [B, 3, npoint, K]
+
+        if self.scorenet_input == 'identity':
+            xyz_features = xyz_diff
+        elif self.scorenet_input == 'neighbor':
+            xyz_features = torch.cat((xyz_diff, points_xyz), dim=1)
+        else:  # ed7
+            euclidian_dist = calc_euclidian_dist(
+                center_xyz.permute(0, 2, 3, 1).reshape(B * npoint * K, 3),
+                points_xyz.permute(0, 2, 3, 1).reshape(B * npoint * K, 3)).\
+                    reshape(B, 1, npoint, K)
+            xyz_features = torch.cat((center_xyz, xyz_diff, euclidian_dist),
+                                     dim=1)
+
+        # scores to assemble weight kernels
+        scores = self.scorenet(xyz_features)  # [B, npoint, K, m]
+
+        # pre-compute features for points and centers separately
+        # features is [B, in_c, N], weight_bank is [C, m * out_dim]
+        point_feat, center_feat = assign_kernel_withoutk(
+            features, self.weight_bank, self.m)
+
+        # aggregate features using custom cuda op
+        new_features = assign_score_cuda(
+            scores, point_feat, center_feat, points_idx,
+            'sum').contiguous()  # [B, out_c, npoint, K]
```

Comment:
Can we merge the general version and CUDA-only version? The main differences, in my opinion, are the additional input of `point_idx` and the way to assign scores/aggregate features?

Reply:
Refactor PAConv to share more functions.